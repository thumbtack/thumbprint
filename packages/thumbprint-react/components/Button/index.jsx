import React from 'react';
import PropTypes from 'prop-types';
import { Themed, Plain } from '../UIAction/index.jsx';

const getCommonProps = props => ({
    onClick: props.onClick,
    isDisabled: props.isDisabled,
    type: props.type,
    children: props.children,
    onMouseEnter: props.onMouseEnter,
    onMouseOver: props.onMouseOver,
    onFocus: props.onFocus,
    onMouseLeave: props.onMouseLeave,
    onBlur: props.onBlur,
    accessibilityLabel: props.accessibilityLabel,
    dataTest: props.dataTest,
});

const TextButton = React.forwardRef((props, ref) => (
    <Plain {...getCommonProps(props)} theme={props.theme} iconLeft={props.iconLeft} ref={ref} />
));

TextButton.propTypes = {
    /**
     * Contents displayed within the button.
     */
    children: PropTypes.node,
    /**
     * Icon from [Thumbprint Icons](/icons/) to render left of the text within `TextButton`.
     */
    iconLeft: PropTypes.node,
    /**
     * Visually and functionally disables the button. We discourage the use of this prop since it
     * is difficult to visually indicate that a link is disabled. Consider not rendering the
     * `TextButton` if it is not interactive.
     */
    isDisabled: PropTypes.bool,
    /**
     * Function that will run when the button is clicked on.
     */
    onClick: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver: PropTypes.func,
    /**
     * Function that runs when the button receives focus.
     */
    onFocus: PropTypes.func,
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave: PropTypes.func,
    /**
     * Function that runs when the button loses focus.
     */
    onBlur: PropTypes.func,
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Controls the button’s background, text, and border color.
     * `inherit` will make the button inherit `color` from its parent.
     */
    theme: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'inherit']),
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type: PropTypes.oneOf(['button', 'submit']),
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest: PropTypes.string,
};

TextButton.defaultProps = {
    children: undefined,
    isDisabled: false,
    iconLeft: null,
    theme: 'primary',
    onClick: null,
    onMouseEnter: undefined,
    onMouseOver: undefined,
    onFocus: undefined,
    onMouseLeave: undefined,
    onBlur: undefined,
    type: 'button',
    accessibilityLabel: undefined,
    dataTest: undefined,
};

const Button = React.forwardRef((props, ref) => (
    <Themed
        {...getCommonProps(props)}
        icon={props.icon}
        isLoading={props.isLoading}
        theme={props.theme}
        size={props.size}
        width={props.width}
        ref={ref}
    />
));

Button.propTypes = {
    /**
     * Contents displayed within the button.
     */
    children: PropTypes.node,
    /**
     * Visually and functionally disables the button.
     */
    isDisabled: PropTypes.bool,
    /**
     * Boolean determining whether the button is in a loading state. When `true` the text will
     * we replaced with a loading animation and interaction will be disabled.
     */
    isLoading: PropTypes.bool,
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the button. It must be one of the
     * `small` icons.
     */
    icon: PropTypes.node,
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type: PropTypes.oneOf(['button', 'submit']),
    /**
     * Function that will run when the button is clicked on.
     */
    onClick: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver: PropTypes.func,
    /**
     * Function that runs when the button receives focus.
     */
    onFocus: PropTypes.func,
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave: PropTypes.func,
    /**
     * Function that runs when the button loses focus.
     */
    onBlur: PropTypes.func,
    /**
     * Controls the button’s background, text, and border color.
     */
    theme: PropTypes.oneOf(['primary', 'secondary', 'caution', 'solid']),
    /**
     * Changes the button's `line-height`, `padding`, `border-radius`, and `font-size`.
     */
    size: PropTypes.oneOf(['small', 'large']),
    /**
     * `Button` components are as wide as the content that is passed in. The `full` option will
     * expand the width to `100%` on all screens. `full-below-small` will expand the width to 100%
     * on devices smaller than [our `small` breakpoint](/tokens/#section-breakpoint).
     */
    width: PropTypes.oneOf(['auto', 'full', 'full-below-small']),
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest: PropTypes.string,
};

Button.defaultProps = {
    children: undefined,
    type: 'button',
    isDisabled: false,
    isLoading: false,
    icon: undefined,
    onClick: undefined,
    onMouseEnter: undefined,
    onMouseOver: undefined,
    onFocus: undefined,
    onMouseLeave: undefined,
    onBlur: undefined,
    theme: 'primary',
    size: 'large',
    width: 'auto',
    accessibilityLabel: undefined,
    dataTest: undefined,
};

export default Button;
export { TextButton };
