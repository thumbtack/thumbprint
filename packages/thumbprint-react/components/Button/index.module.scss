@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';
@import '~@thumbtack/thumbprint-scss/mixins';

$border-width: 2px;

.root {
    border: 2px solid;
    border-radius: 4px;
    font-weight: $tp-font-weight__bold;
    // The designs have an inset 2px border. The subtraction accounts for that.
    padding: 0 ($tp-space__4 - $border-width);
    // Used to absolutely position the `LoaderDots`
    position: relative;
    display: inline-block;
    user-select: none;
}

.buttonWidthFull {
    width: 100%;
}

.buttonWidthFullBelowSmall {
    @include tp-respond-below($tp-breakpoint__small) {
        width: 100%;
    }
}

.buttonLarge,
.buttonMedium {
    min-height: (52px - ($border-width * 2));
}

.buttonLarge {
    // This doesn't use one of our system font sizes because the text size needs to be just large
    // enough so that it is accessible on our blue background.
    font-size: 19px;
}

.buttonMedium {
    font-size: $tp-font__body__1__size;

    @include tp-respond-below($tp-breakpoint__small) {
        min-height: 48px;
    }
}

.buttonSmall {
    font-size: $tp-font__body__2__size;
    min-height: 36px - ($border-width * 2);
}

.primaryButton {
    color: $tp-color__white;
    border-color: $tp-color__blue;
    background-color: $tp-color__blue;

    &[disabled] {
        border-color: $tp-color__blue-300;
        background-color: $tp-color__blue-300;
    }
}

.secondaryButtonEmphasisHigh {
    color: $tp-color__white;
    border-color: $tp-color__black;
    background-color: $tp-color__black;
}

.secondaryButtonEmphasisLow {
    color: $tp-color__black;
    border-color: $tp-color__white;
    background-color: $tp-color__white;
}

.secondaryButtonBlackBorder {
    border-color: $tp-color__black;

    &[disabled] {
        background-color: $tp-color__gray;
        border-color: $tp-color__gray;
    }
}

.secondaryButtonWhiteBorder {
    border-color: $tp-color__white;

    &[disabled] {
        color: $tp-color__gray;
    }
}

.secondaryButtonGrayBorder {
    border-color: $tp-color__gray;

    &[disabled] {
        color: $tp-color__gray;
    }
}

.tertiaryButton {
    border-color: $tp-color__white;
    background-color: $tp-color__white;
    color: $tp-color__black-300;
}

.cautionButton[disabled] {
    border-color: $tp-color__gray;
    background-color: $tp-color__red-100;
    color: $tp-color__red-200;
}

.cautionButtonEmphasisHigh {
    background-color: $tp-color__red-500;
    color: $tp-color__white;
    border-color: $tp-color__red-500;
}

.cautionButtonEmphasisLow {
    background-color: $tp-color__white;
    color: $tp-color__red-500;
    border-color: $tp-color__gray;
}

.flexWrapper {
    // This wraps the content since `button`s are not able to use `display: flex`.
    display: flex;
    align-items: center;
    justify-content: center;
}
