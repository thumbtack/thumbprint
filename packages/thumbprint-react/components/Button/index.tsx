import React from 'react';
import { Themed, Plain } from '../UIAction/index';

interface CommonProps {
    children?: React.ReactNode;
    iconLeft?: React.ReactNode;
    iconRight?: React.ReactNode;
    isDisabled?: boolean;
    onClick?: () => void;
    onMouseEnter?: () => void;
    onMouseOver?: () => void;
    onFocus?: () => void;
    onMouseLeave?: () => void;
    onBlur?: () => void;
    theme?: 'primary' | 'secondary' | 'tertiary' | 'inherit';
    type?: 'button' | 'submit';
    dataTest?: string;
    accessibilityLabel?: string;
}

const getCommonProps = (props: CommonProps): CommonProps => ({
    onClick: props.onClick,
    isDisabled: props.isDisabled,
    type: props.type,
    children: props.children,
    onMouseEnter: props.onMouseEnter,
    onMouseOver: props.onMouseOver,
    onFocus: props.onFocus,
    onMouseLeave: props.onMouseLeave,
    onBlur: props.onBlur,
    accessibilityLabel: props.accessibilityLabel,
    dataTest: props.dataTest,
});

const TextButton = React.forwardRef<HTMLButtonElement, TextButtonPropTypes>(
    (
        {
            children,
            iconLeft,
            iconRight,
            isDisabled = false,
            onClick,
            onMouseEnter,
            onMouseOver,
            onFocus,
            onMouseLeave,
            onBlur,
            accessibilityLabel,
            theme = 'primary',
            type = 'button',
            dataTest,
        }: TextButtonPropTypes,
        ref,
    ): JSX.Element => (
        <Plain
            {...getCommonProps({
                onClick,
                isDisabled,
                type,
                children,
                onMouseEnter,
                onMouseOver,
                onFocus,
                onMouseLeave,
                onBlur,
                accessibilityLabel,
                dataTest,
            })}
            theme={theme}
            iconLeft={iconLeft}
            iconRight={iconRight}
            ref={ref}
        />
    ),
);

interface TextButtonPropTypes {
    /**
     * Contents displayed within the button.
     */
    children?: React.ReactNode;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render left of the text within `TextButton`.
     */
    iconLeft?: React.ReactNode;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render right of the text within `TextButton`.
     */
    iconRight?: React.ReactNode;
    /**
     * Visually and functionally disables the button. We discourage the use of this prop since it
     * is difficult to visually indicate that a link is disabled. Consider not rendering the
     * `TextButton` if it is not interactive.
     */
    isDisabled?: boolean;
    /**
     * Function that will run when the button is clicked on.
     */
    onClick?: () => void;
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter?: () => void;
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver?: () => void;
    /**
     * Function that runs when the button receives focus.
     */
    onFocus?: () => void;
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave?: () => void;
    /**
     * Function that runs when the button loses focus.
     */
    onBlur?: () => void;
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel?: string;
    /**
     * Controls the button’s background, text, and border color.
     * `inherit` will make the button inherit `color` from its parent.
     */
    theme?: 'primary' | 'secondary' | 'tertiary' | 'inherit';
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type?: 'button' | 'submit';
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonPropTypes>(
    (
        {
            children,
            isDisabled = false,
            isLoading = false,
            icon,
            accessibilityLabel,
            type = 'button',
            onClick,
            onMouseEnter,
            onMouseOver,
            onFocus,
            onMouseLeave,
            onBlur,
            theme = 'primary',
            size = 'large',
            width = 'auto',
            dataTest,
        }: ButtonPropTypes,
        ref,
    ): JSX.Element => (
        <Themed
            {...getCommonProps({
                onClick,
                isDisabled,
                type,
                children,
                onMouseEnter,
                onMouseOver,
                onFocus,
                onMouseLeave,
                onBlur,
                accessibilityLabel,
                dataTest,
            })}
            icon={icon}
            isLoading={isLoading}
            theme={theme}
            size={size}
            width={width}
            ref={ref}
        />
    ),
);

interface ButtonPropTypes {
    /**
     * Contents displayed within the button.
     */
    children?: React.ReactNode;
    /**
     * Visually and functionally disables the button.
     */
    isDisabled?: boolean;
    /**
     * Boolean determining whether the button is in a loading state. When `true` the text will
     * we replaced with a loading animation and interaction will be disabled.
     */
    isLoading?: boolean;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the button. It must be one of the
     * `small` icons.
     */
    icon?: React.ReactNode;
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel?: string;
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type?: 'button' | 'submit';
    /**
     * Function that will run when the button is clicked on.
     */
    onClick?: () => void;
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter?: () => void;
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver?: () => void;
    /**
     * Function that runs when the button receives focus.
     */
    onFocus?: () => void;
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave?: () => void;
    /**
     * Function that runs when the button loses focus.
     */
    onBlur?: () => void;
    /**
     * Controls the button’s background, text, and border color.
     */
    theme?: 'primary' | 'secondary' | 'tertiary' | 'caution' | 'solid';
    /**
     * Changes the button's `line-height`, `padding`, `border-radius`, and `font-size`.
     */
    size?: 'small' | 'large';
    /**
     * `Button` components are as wide as the content that is passed in. The `full` option will
     * expand the width to `100%` on all screens. `full-below-small` will expand the width to 100%
     * on devices smaller than [our `small` breakpoint](/tokens/#section-breakpoint).
     */
    width?: 'auto' | 'full' | 'full-below-small';
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
}

export default Button;
export { TextButton };
