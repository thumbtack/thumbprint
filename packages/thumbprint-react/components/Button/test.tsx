import React from 'react';
import { mount } from 'enzyme';
import Button, { TextButton } from './index';

test('renders children passed in', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component>Goose</Component>);
        expect(wrapper.text()).toEqual('Goose');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders button of type `button` by default', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component>Goose</Component>);
        expect(wrapper.find('button').prop('type')).toEqual('button');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders button of type `submit` when `type` is set', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component type="submit">Goose</Component>);
        expect(wrapper.find('button').prop('type')).toEqual('submit');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('`onClick` runs when button is clicked on', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const onClick = jest.fn();
        const wrapper = mount(<Component onClick={onClick}>Goose</Component>);

        expect(onClick).toHaveBeenCalledTimes(0);
        wrapper.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('`onMouseOver` runs when button is moused over', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const onMouseOver = jest.fn();
        // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
        const wrapper = mount(<Component onMouseOver={onMouseOver}>Goose</Component>);

        expect(onMouseOver).toHaveBeenCalledTimes(0);
        wrapper.simulate('mouseover');
        expect(onMouseOver).toHaveBeenCalledTimes(1);
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders icon that is passed in', (): void => {
    const wrapper = mount(<Button icon={<svg>Duck</svg>}>Goose</Button>);
    expect(wrapper.find('svg')).toHaveLength(1);
    expect(wrapper.text()).toBe('DuckGoose');
    expect(wrapper).toMatchSnapshot();
});

test('adds `dataTest` prop', (): void => {
    const wrapperA = mount(<Button dataTest="Duck">Goose</Button>);
    expect(wrapperA.find('button').prop('data-test')).toEqual('Duck');
    expect(wrapperA).toMatchSnapshot();

    const wrapperB = mount(<TextButton dataTest="Duck">Goose</TextButton>);
    expect(wrapperB.find('button').prop('data-test')).toEqual('Duck');
    expect(wrapperB).toMatchSnapshot();
});
