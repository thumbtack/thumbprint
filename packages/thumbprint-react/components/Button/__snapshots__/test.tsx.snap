// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<ForwardRef
  dataTest="Duck"
>
  <ForwardRef
    dataTest="Duck"
    isDisabled={false}
    isLoading={false}
    size="large"
    theme="primary"
    type="button"
    width="auto"
  >
    <button
      className="themedButton themedButtonRoundedBordersLeft themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthAuto"
      data-test="Duck"
      disabled={false}
      type="button"
    >
      <span
        className="flexWrapper flexWrapperSizeLarge"
      >
        Goose
      </span>
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`adds \`dataTest\` prop 2`] = `
<ForwardRef
  dataTest="Duck"
>
  <ForwardRef
    dataTest="Duck"
    isDisabled={false}
    theme="primary"
    type="button"
  >
    <button
      className="plain plainThemePrimary"
      data-test="Duck"
      disabled={false}
      type="button"
    >
      Goose
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders button of type \`button\` by default 1`] = `
<ForwardRef>
  <ForwardRef
    isDisabled={false}
    isLoading={false}
    size="large"
    theme="primary"
    type="button"
    width="auto"
  >
    <button
      className="themedButton themedButtonRoundedBordersLeft themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthAuto"
      disabled={false}
      type="button"
    >
      <span
        className="flexWrapper flexWrapperSizeLarge"
      >
        Goose
      </span>
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders button of type \`button\` by default 2`] = `
<ForwardRef>
  <ForwardRef
    isDisabled={false}
    theme="primary"
    type="button"
  >
    <button
      className="plain plainThemePrimary"
      disabled={false}
      type="button"
    >
      Goose
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders button of type \`submit\` when \`type\` is set 1`] = `
<ForwardRef
  type="submit"
>
  <ForwardRef
    isDisabled={false}
    isLoading={false}
    size="large"
    theme="primary"
    type="submit"
    width="auto"
  >
    <button
      className="themedButton themedButtonRoundedBordersLeft themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthAuto"
      disabled={false}
      type="submit"
    >
      <span
        className="flexWrapper flexWrapperSizeLarge"
      >
        Goose
      </span>
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders button of type \`submit\` when \`type\` is set 2`] = `
<ForwardRef
  type="submit"
>
  <ForwardRef
    isDisabled={false}
    theme="primary"
    type="submit"
  >
    <button
      className="plain plainThemePrimary"
      disabled={false}
      type="submit"
    >
      Goose
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders children passed in 1`] = `
<ForwardRef>
  <ForwardRef
    isDisabled={false}
    isLoading={false}
    size="large"
    theme="primary"
    type="button"
    width="auto"
  >
    <button
      className="themedButton themedButtonRoundedBordersLeft themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthAuto"
      disabled={false}
      type="button"
    >
      <span
        className="flexWrapper flexWrapperSizeLarge"
      >
        Goose
      </span>
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders children passed in 2`] = `
<ForwardRef>
  <ForwardRef
    isDisabled={false}
    theme="primary"
    type="button"
  >
    <button
      className="plain plainThemePrimary"
      disabled={false}
      type="button"
    >
      Goose
    </button>
  </ForwardRef>
</ForwardRef>
`;

exports[`renders icon that is passed in 1`] = `
<ForwardRef
  icon={
    <svg>
      Duck
    </svg>
  }
>
  <ForwardRef
    icon={
      <svg>
        Duck
      </svg>
    }
    isDisabled={false}
    isLoading={false}
    size="large"
    theme="primary"
    type="button"
    width="auto"
  >
    <button
      className="themedButton themedButtonRoundedBordersLeft themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthAuto"
      disabled={false}
      type="button"
    >
      <span
        className="flexWrapper flexWrapperSizeLarge"
      >
        <span
          className="flexCenter"
        >
          <span
            className="iconContainer iconContainerHasChildren"
          >
            <svg>
              Duck
            </svg>
          </span>
          Goose
        </span>
      </span>
    </button>
  </ForwardRef>
</ForwardRef>
`;
